demos/coreHTTP/http_demo_s3_download_multithreaded.c

```diff
@@ -0,0 +1,965 @@
+#ifndef democonfigS3_PRESIGNED_GET_URL
+    #define democonfigS3_PRESIGNED_GET_URL    "GET-URL"
+#endif
+
+/* Check that a timeout for transport send and receive functions is defined.*/
+#ifndef democonfigTRANSPORT_SEND_RECV_TIMEOUT_MS
+    #define democonfigTRANSPORT_SEND_RECV_TIMEOUT_MS    ( 5000 )
+#endif
+
+/* Check that size of the user buffer is defined. */
+#ifndef democonfigUSER_BUFFER_LENGTH
+    #define democonfigUSER_BUFFER_LENGTH    ( 4096 )
+#endif
+
+/* Check that the range request length is defined. */
+#ifndef democonfigRANGE_REQUEST_LENGTH
+    #define democonfigRANGE_REQUEST_LENGTH    ( 2048 )
+#endif
+
+/**
+ * @brief The maximum number of times to run the loop in this demo.
+ */
+#ifndef httpexampleMAX_DEMO_COUNT
+    #define httpexampleMAX_DEMO_COUNT    ( 3 )
+#endif
+
+/**
+ * @brief Length of the pre-signed GET URL defined in http_demo_s3_download_multithreaded_config.h.
+ */
+#define httpexampleS3_PRESIGNED_GET_URL_LENGTH               ( sizeof( democonfigS3_PRESIGNED_GET_URL ) - 1 )
+
+/**
+ * @brief The length of the HTTP GET method.
+ */
+#define httpexampleHTTP_METHOD_GET_LENGTH                    ( sizeof( HTTP_METHOD_GET ) - 1 )
+
+/**
+ * @brief Field name of the HTTP range header to read from server response.
+ */
+#define httpexampleHTTP_CONTENT_RANGE_HEADER_FIELD           "Content-Range"
+
+/**
+ * @brief Length of the HTTP range header field.
+ */
+#define httpexampleHTTP_CONTENT_RANGE_HEADER_FIELD_LENGTH    ( sizeof( httpexampleHTTP_CONTENT_RANGE_HEADER_FIELD ) - 1 )
+
+/**
+ * @brief The HTTP status code returned for partial content.
+ */
+#define httpexampleHTTP_STATUS_CODE_PARTIAL_CONTENT          206
+
+
+/**
+ * @brief Time in ticks to wait between each iteration of
+ * RunCoreHttpS3DownloadMultithreadedDemo().
+ */
+#define httpexampleDELAY_BETWEEN_DEMO_ITERATIONS_TICKS    ( pdMS_TO_TICKS( 5000U ) )
+
+/**
+ * @brief The stack size to use for HTTP tasks.
+ */
+#define httpexampleTASK_STACK_SIZE                        ( configMINIMAL_STACK_SIZE * 2 )
+
+/**
+ * @brief Ticks to wait for task notifications.
+ */
+#define httpexampleDEMO_TICKS_TO_WAIT                     pdMS_TO_TICKS( 1000 )
+
+/**
+ * @brief Represents the network context used for the TLS session with the
+ * server.
+ */
+static NetworkContext_t xNetworkContext;
+
+/**
+ * @brief The host address string extracted from the pre-signed URL.
+ *
+ * @note httpexampleS3_PRESIGNED_GET_URL_LENGTH is set as the array length here
+ * as the length of the host name string cannot exceed this value.
+ */
+static char cServerHost[ httpexampleS3_PRESIGNED_GET_URL_LENGTH ];
+
+/**
+ * @brief The length of the host address found in the pre-signed URL.
+ */
+static size_t xServerHostLength;
+
+/**
+ * @brief Data type for request queue.
+ *
+ * In addition to sending #HTTPRequestHeaders_t, we need to send the buffer it
+ * uses as the pointer to the buffer in #HTTPRequestHeaders_t may not be
+ * accurate after being copied into the queue.
```

Comment:
This is not clear to me. Did you meant that the buffer is copied to ucHeaderBuffer?

Reply:
Updated the comment to be less misleading -- the buffer itself isn't copied, only the pointers are updated.